### PR TITLE
fix: TOCTOU race in lock acquisition and add fileno() to _QueueWriter

### LINKED ISSUE #24: Race condition in lock file PID check (TOCTOU)
## Description
There's a time-of-check-time-of-use (TOCTOU) race condition in the lock file implementation's PID checking logic.

## Location
`src/pivot/lock.py` - lock acquisition code

## Problem
The code:
1. Reads the PID from a stale lock file
2. Checks if that process is alive using `os.kill(pid, 0)`
3. If dead, removes the lock and acquires it

Between steps 2 and 3, another process could:
- Reuse that PID (PID recycling)
- Acquire the lock legitimately

This creates a window where two processes believe they hold the lock.

## Impact
- **Severity**: High in high-concurrency environments
- **Likelihood**: Low on most systems (requires PID recycling during the race window)
- Could lead to concurrent execution of stages that should be mutually exclusive

## References
- Classic TOCTOU vulnerability pattern
- Similar to lock file handling in package managers

---

### 1. Problem Analysis

The current implementation in `/home/pivot/app/src/pivot/executor/worker.py` (lines 314-345) has a TOCTOU (Time-Of-Check-Time-Of-Use) race condition:

```python
def _acquire_execution_lock(stage_name: str, cache_dir: pathlib.Path) -> pathlib.Path:
    cache_dir.mkdir(parents=True, exist_ok=True)
    sentinel = cache_dir / f"{stage_name}.running"

    for _ in range(_MAX_LOCK_ATTEMPTS):
        try:
            fd = os.open(sentinel, os.O_CREAT | os.O_EXCL | os.O_WRONLY)  # Step 1: Try atomic create
            os.write(fd, f"pid: {os.getpid()}\n".encode())
            os.close(fd)
            return sentinel
        except FileExistsError:
            pass

        try:
            content = sentinel.read_text()                               # Step 2: Read PID
            pid = int(content.split(":")[1].strip())
        except (ValueError, IndexError, OSError):
            pid = None

        if pid is not None and pid > 0 and _is_process_alive(pid):       # Step 3: Check if alive
            raise exceptions.StageAlreadyRunningError(...)

        sentinel.unlink(missing_ok=True)                                  # Step 4: Delete stale lock
        # RACE WINDOW: Between unlink and next O_CREAT | O_EXCL
        # - Another process could create the lock legitimately
        # - PID could be recycled, and new process reuses exact same PID
```

**Race Condition Scenario:**

1. Process A reads PID `12345` from lock file, sees it's dead
2. Process B (a new process) is assigned PID `12345` by the OS (PID recycling)
3. Process A deletes the lock file  
4. Process B atomically creates the lock file with PID `12345`
5. Process A's next iteration creates a new lock file
6. Both processes now believe they hold the lock

### 2. Solution Design

The fix uses **atomic rename for lock takeover** instead of delete-then-create. This eliminates the race window by making the "remove stale + acquire" operation atomic.

**Key Principle:** Instead of `unlink()` then retry `O_CREAT | O_EXCL`, we create our candidate lock file first, then atomically `rename()` it over the stale lock.

**New Algorithm:**

```
1. Try atomic create with O_CREAT | O_EXCL
2. If fails (FileExistsError):
   a. Read the existing lock's PID
   b. Check if that process is alive
   c. If alive: raise StageAlreadyRunningError
   d. If dead (stale lock):
      i.   Create a NEW temp file with our PID
      ii.  Atomic rename temp -> sentinel (overwrites stale lock)
      iii. Verify we still own the lock (re-read and check PID)
      iv.  If verification fails, another process won - retry
```

The verification step (step 2.d.iii) is crucial: after the rename, we re-read the file to confirm our PID is there. This handles the case where another process did the same rename at the same moment.

#### 3.1 Modified `_acquire_execution_lock` Function

Location: `/home/pivot/app/src/pivot/executor/worker.py`, lines 314-345

```python
def _acquire_execution_lock(stage_name: str, cache_dir: pathlib.Path) -> pathlib.Path:
    """Acquire exclusive lock for stage execution using atomic operations."""
    cache_dir.mkdir(parents=True, exist_ok=True)
    sentinel = cache_dir / f"{stage_name}.running"

    for attempt in range(_MAX_LOCK_ATTEMPTS):
        # Fast path: try atomic create
        try:
            fd = os.open(sentinel, os.O_CREAT | os.O_EXCL | os.O_WRONLY)
            os.write(fd, f"pid: {os.getpid()}\n".encode())
            os.close(fd)
            return sentinel
        except FileExistsError:
            pass

        # Lock exists - check if it's stale
        existing_pid = _read_lock_pid(sentinel)
        
        if existing_pid is not None and _is_process_alive(existing_pid):
            raise exceptions.StageAlreadyRunningError(
                f"Stage '{stage_name}' is already running (PID {existing_pid})"
            )

        # Stale lock detected - attempt atomic takeover
        if _atomic_lock_takeover(sentinel, existing_pid):
            return sentinel
        # Takeover failed (another process beat us) - retry loop

    raise exceptions.StageAlreadyRunningError(
        f"Failed to acquire lock for '{stage_name}' after {_MAX_LOCK_ATTEMPTS} attempts"
    )
```

#### 3.2 New Helper: `_read_lock_pid`

```python
def _read_lock_pid(sentinel: pathlib.Path) -> int | None:
    """Read PID from lock file. Returns None if file is missing, corrupted, or has invalid PID."""
    try:
        content = sentinel.read_text()
        pid = int(content.split(":")[1].strip())
        return pid if pid > 0 else None
    except (FileNotFoundError, ValueError, IndexError, OSError):
        return None
```

#### 3.3 New Helper: `_atomic_lock_takeover`

```python
def _atomic_lock_takeover(sentinel: pathlib.Path, stale_pid: int | None) -> bool:
    """Atomically take over a stale lock file.
    
    Uses temp file + rename pattern to eliminate TOCTOU race.
    Returns True if we successfully acquired the lock, False otherwise.
    """
    my_pid = os.getpid()
    
    # Create temp file with our PID (same directory for atomic rename)
    fd, tmp_path = tempfile.mkstemp(dir=sentinel.parent, prefix=f".{sentinel.name}.")
    try:
        os.write(fd, f"pid: {my_pid}\n".encode())
        os.close(fd)
        
        # Atomic rename: overwrites existing file
        os.rename(tmp_path, sentinel)
        
        # Verify we still hold the lock (another process may have done the same)
        verified_pid = _read_lock_pid(sentinel)
        if verified_pid == my_pid:
            if stale_pid is not None:
                logger.warning(f"Removed stale lock file: {sentinel} (was PID {stale_pid})")
            return True
        
        # Another process beat us to it
        return False
        
    except OSError:
        # Clean up temp file on any error
        with contextlib.suppress(OSError):
            os.unlink(tmp_path)
        return False
```

#### 3.4 Import Addition

Add `tempfile` to imports at the top of the file:

```python
import tempfile
```

### 4. Why This Solution Works

1. **No delete-then-create gap:** The stale lock is never deleted; it's atomically replaced by rename.

2. **PID recycling protection:** Even if a new process gets the same PID as the dead one:
   - The verification step (`_read_lock_pid` after rename) ensures we actually own the lock
   - If two processes race to take over, only one can succeed (the one whose PID ends up in the file)

3. **Atomic rename guarantees:** On POSIX systems, `rename()` is atomic - the target file either has the old content or the new content, never an intermediate state.

4. **Maintains existing behavior:** 
   - Live processes still block acquisition
   - Corrupted lock files are treated as stale
   - Negative/invalid PIDs are treated as stale

### 5. Edge Cases Handled

| Scenario | Current (Buggy) | New (Fixed) |
|----------|-----------------|-------------|
| Two processes race to take over stale lock | Both may succeed (race) | Exactly one succeeds |
| PID recycling during takeover | May cause double-lock | Verified by re-read |
| Process dies after creating temp file | Orphaned temp (unlikely) | Cleaned up in except block |
| Lock file deleted externally during takeover | Next iteration succeeds | Same behavior |
| Disk full during temp file write | OSError, retry | Same, explicit cleanup |

#### 6.1 Existing Tests (Should Still Pass)

All existing tests in `/home/pivot/app/tests/execution/test_executor_worker.py` (lines 598-710) should continue to pass:

- `test_execution_lock_creates_sentinel_file`
- `test_execution_lock_removes_sentinel_on_exception`
- `test_acquire_execution_lock_succeeds_when_available`
- `test_acquire_execution_lock_fails_when_held_by_live_process`
- `test_acquire_execution_lock_breaks_stale_lock`
- `test_acquire_execution_lock_breaks_corrupted_lock`
- `test_acquire_execution_lock_breaks_negative_pid_lock`
- `test_is_process_alive_*`

#### 6.2 New Tests to Add

```python
def test_atomic_lock_takeover_creates_lock_with_own_pid(worker_env: pathlib.Path) -> None:
    """Atomic takeover creates lock with current process PID."""
    sentinel = worker_env / "test_stage.running"
    sentinel.write_text("pid: 999999999\n")  # Stale lock
    
    result = worker._atomic_lock_takeover(sentinel, 999999999)
    
    assert result is True
    assert sentinel.exists()
    content = sentinel.read_text()
    assert f"pid: {os.getpid()}" in content
    sentinel.unlink()


def test_atomic_lock_takeover_fails_when_another_process_wins(
    worker_env: pathlib.Path,
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Atomic takeover returns False when another process beats us."""
    sentinel = worker_env / "test_stage.running"
    original_rename = os.rename
    
    def sneaky_rename(src: str, dst: str) -> None:
        """Simulate another process winning the race after our rename."""
        original_rename(src, dst)
        # Immediately overwrite with different PID to simulate race
        pathlib.Path(dst).write_text("pid: 999888777\n")
    
    monkeypatch.setattr(os, "rename", sneaky_rename)
    sentinel.write_text("pid: 999999999\n")  # Stale lock
    
    result = worker._atomic_lock_takeover(sentinel, 999999999)
    
    assert result is False
    sentinel.unlink()


def test_acquire_lock_retries_after_failed_takeover(
    worker_env: pathlib.Path,
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Lock acquisition retries when atomic takeover fails."""
    sentinel = worker_env / "test_stage.running"
    call_count = 0
    
    def mock_takeover(sent: pathlib.Path, pid: int | None) -> bool:
        nonlocal call_count
        call_count += 1
        if call_count < 2:
            return False  # Fail first attempt
        # Second attempt: actually create the lock
        sent.write_text(f"pid: {os.getpid()}\n")
        return True
    
    # Start with stale lock
    sentinel.write_text("pid: 999999999\n")
    monkeypatch.setattr(worker, "_atomic_lock_takeover", mock_takeover)
    
    result = worker._acquire_execution_lock("test_stage", worker_env)
    
    assert result == sentinel
    assert call_count >= 2
    sentinel.unlink()


def test_read_lock_pid_returns_pid_for_valid_file(worker_env: pathlib.Path) -> None:
    """_read_lock_pid extracts PID from valid lock file."""
    sentinel = worker_env / "test.running"
    sentinel.write_text("pid: 12345\n")
    
    assert worker._read_lock_pid(sentinel) == 12345
    sentinel.unlink()


def test_read_lock_pid_returns_none_for_missing_file(worker_env: pathlib.Path) -> None:
    """_read_lock_pid returns None for non-existent file."""
    sentinel = worker_env / "nonexistent.running"
    assert worker._read_lock_pid(sentinel) is None


def test_read_lock_pid_returns_none_for_corrupted_file(worker_env: pathlib.Path) -> None:
    """_read_lock_pid returns None for corrupted content."""
    sentinel = worker_env / "test.running"
    sentinel.write_text("garbage content")
    
    assert worker._read_lock_pid(sentinel) is None
    sentinel.unlink()


def test_read_lock_pid_returns_none_for_negative_pid(worker_env: pathlib.Path) -> None:
    """_read_lock_pid returns None for invalid negative PID."""
    sentinel = worker_env / "test.running"
    sentinel.write_text("pid: -1\n")
    
    assert worker._read_lock_pid(sentinel) is None
    sentinel.unlink()


def test_atomic_takeover_cleans_temp_on_error(
    worker_env: pathlib.Path,
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Temp file is cleaned up when rename fails."""
    sentinel = worker_env / "test_stage.running"
    sentinel.write_text("pid: 999999999\n")
    
    def failing_rename(src: str, dst: str) -> None:
        raise OSError("Simulated disk error")
    
    monkeypatch.setattr(os, "rename", failing_rename)
    
    result = worker._atomic_lock_takeover(sentinel, 999999999)
    
    assert result is False
    # Verify no temp files left behind
    temp_files = list(worker_env.glob(".test_stage.running.*"))
    assert len(temp_files) == 0
    sentinel.unlink()
```

### 7. Implementation Sequence

1. **Add import:** Add `import tempfile` to imports
2. **Add `_read_lock_pid` helper:** Extract PID reading logic into separate function
3. **Add `_atomic_lock_takeover` helper:** Implement atomic takeover logic
4. **Modify `_acquire_execution_lock`:** Use new helpers, remove `sentinel.unlink()`
5. **Add tests:** Unit tests for new helpers, integration test for race condition
6. **Run validation:** `ruff format .`, `ruff check .`, `basedpyright .`, `pytest tests/`

### 8. Potential Issues and Mitigations

**Issue 1: Temp file cleanup on crash**
- Mitigation: Temp files use predictable prefix (`.{stage_name}.running.`) so they can be identified and cleaned up by future runs or a cleanup script.

**Issue 2: Temp file in same directory requirement**
- Mitigation: `tempfile.mkstemp(dir=sentinel.parent)` ensures temp file is on same filesystem as target, required for atomic rename.

**Issue 3: Windows compatibility**
- Note: `os.rename()` on Windows cannot overwrite existing files. This is a Linux/POSIX-specific fix. For Windows, need `os.replace()` instead. Consider using `os.replace()` which works on both platforms.

### Critical Files for Implementation

- `/home/pivot/app/src/pivot/executor/worker.py` - Contains `_acquire_execution_lock` function with the TOCTOU bug (lines 314-345)
- `/home/pivot/app/tests/execution/test_executor_worker.py` - Existing tests for execution lock (lines 598-710) to extend
- `/home/pivot/app/src/pivot/cache.py` - Reference for atomic write patterns using temp file + rename (lines 29-58)
- `/home/pivot/app/src/pivot/exceptions.py` - `StageAlreadyRunningError` exception class (lines 55-58)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### LINKED ISSUE #26: _QueueWriter missing fileno() method causes AttributeError with some libraries
## Description
The `_QueueWriter` class used to capture stdout/stderr in worker processes is missing a `fileno()` method, which some libraries expect on file-like objects.

## Location
`src/pivot/executor/worker.py` - `_QueueWriter` class

## Problem
`_QueueWriter` implements a minimal file-like interface for capturing output, but omits `fileno()`. Some libraries and tools:
- Check `hasattr(sys.stdout, 'fileno')` before calling it
- Call `fileno()` directly expecting it to exist on stdout
- Use `fileno()` to detect if output is a TTY

When these libraries are used in a stage function, they get `AttributeError: '_QueueWriter' object has no attribute 'fileno'`.

## Affected Libraries (examples)
- `tqdm` progress bars
- `rich` console library
- Some logging handlers
- Libraries that check `isatty()`

## Workaround
Users can catch the exception in their stage functions or configure libraries to not use TTY features.

---

# Implementation Plan: Add `fileno()` method to `_QueueWriter`

## Problem Summary

The `_QueueWriter` class in `/home/pivot/app/src/pivot/executor/worker.py` implements a minimal file-like interface for capturing stdout/stderr in worker processes. It currently lacks a `fileno()` method, which causes `AttributeError` when libraries like `tqdm`, `rich`, or logging handlers attempt to call `stdout.fileno()` or `stderr.fileno()`.

## Current State

The `_QueueWriter` class (lines 233-301) currently implements:
- `__init__()` - Initialize with stage name, queue, stderr flag, and output lines list
- `__enter__()` / `__exit__()` - Context manager for stream redirection
- `write(s: str) -> int` - Buffers and sends lines to queue
- `flush()` - Flushes any remaining buffered content
- `isatty() -> bool` - Returns `False` (already implemented at line 300-301)

Missing: `fileno()` method

### Approach

Follow Python's standard library conventions as demonstrated by `io.StringIO`:

1. **Add `fileno()` method** that raises `io.UnsupportedOperation` with a clear message
2. **Import `io` module** following project's Google-style import conventions

This approach is preferred over:
- **Omitting the method**: Some libraries call `fileno()` directly without checking `hasattr()` first
- **Raising `OSError`**: Less specific than `io.UnsupportedOperation`
- **Raising `ValueError`**: The pytest project discussed moving away from this toward `io.UnsupportedOperation`

### Why `io.UnsupportedOperation`?

1. **Python standard**: `io.StringIO().fileno()` raises this exception
2. **Backward compatible**: Inherits from both `OSError` and `ValueError`, so code catching either will still work
3. **Explicit**: The exception type clearly communicates "unsupported operation" rather than a generic error

### Library Compatibility Analysis

Common library behaviors:

| Library | Check Pattern | Result with our fix |
|---------|--------------|---------------------|
| **tqdm** | Uses `hasattr(file, "isatty")` and calls `file.isatty()` | Already works (has `isatty()`) |
| **rich** | Uses `getattr(file, "isatty", None)` and catches exceptions | Already works, `fileno()` gracefully handled |
| **logging** | Some handlers call `stream.fileno()` | Will raise clear exception instead of `AttributeError` |
| **os.isatty()** | Calls `fileno()` on stream | Will raise clear exception |

### Step 1: Add `io` import

At line 9, add `import io` in alphabetical order with other stdlib imports:

```python
import contextlib
import io  # ADD THIS
import logging
import os
import pathlib
import queue
```

### Step 2: Add `fileno()` method to `_QueueWriter`

Add after the existing `isatty()` method (after line 301):

```python
def fileno(self) -> int:
    """Raise UnsupportedOperation - _QueueWriter has no underlying file descriptor."""
    raise io.UnsupportedOperation("_QueueWriter does not use a file descriptor")
```

### Complete `_QueueWriter` class after changes

The class would look like (showing relevant methods):

```python
class _QueueWriter:
    """Context manager for capturing stdout/stderr to a queue.

    Handles stream redirection, output capture, and automatic flushing.
    Implements minimal file-like interface needed by print() and common libraries.
    """

    # ... existing attributes and methods ...

    def flush(self) -> None:
        if self._buffer:
            self._send_line(self._buffer)
            self._buffer = ""

    def isatty(self) -> bool:
        return False

    def fileno(self) -> int:
        """Raise UnsupportedOperation - _QueueWriter has no underlying file descriptor."""
        raise io.UnsupportedOperation("_QueueWriter does not use a file descriptor")
```

### Test 1: `fileno()` raises correct exception

Add to `/home/pivot/app/tests/execution/test_executor_worker.py` in the `_QueueWriter Tests` section (after line 530):

```python
def test_queue_writer_fileno_raises_unsupported_operation() -> None:
    """_QueueWriter.fileno() raises io.UnsupportedOperation."""
    import io
    
    output_lines: list[tuple[str, bool]] = []
    queue_obj: mp.Queue[OutputMessage] = mp.Manager().Queue()  # pyright: ignore[reportAssignmentType]

    writer = worker._QueueWriter(
        "test_stage",
        queue_obj,
        is_stderr=False,
        output_lines=output_lines,
    )

    with pytest.raises(io.UnsupportedOperation, match="file descriptor"):
        writer.fileno()
```

### Test 2: Exception inherits from expected base classes

```python
def test_queue_writer_fileno_exception_inheritance() -> None:
    """_QueueWriter.fileno() exception inherits from OSError and ValueError."""
    import io
    
    output_lines: list[tuple[str, bool]] = []
    queue_obj: mp.Queue[OutputMessage] = mp.Manager().Queue()  # pyright: ignore[reportAssignmentType]

    writer = worker._QueueWriter(
        "test_stage",
        queue_obj,
        is_stderr=False,
        output_lines=output_lines,
    )

    try:
        writer.fileno()
    except io.UnsupportedOperation as e:
        # Verify inheritance chain for backward compatibility
        assert isinstance(e, OSError), "Should inherit from OSError"
        assert isinstance(e, ValueError), "Should inherit from ValueError"
```

### Test 3: Integration test with library that uses `fileno()`

```python
def test_queue_writer_with_os_isatty_check() -> None:
    """Libraries using os.isatty(stream.fileno()) get clear error, not AttributeError."""
    import io
    
    output_lines: list[tuple[str, bool]] = []
    queue_obj: mp.Queue[OutputMessage] = mp.Manager().Queue()  # pyright: ignore[reportAssignmentType]

    writer = worker._QueueWriter(
        "test_stage",
        queue_obj,
        is_stderr=False,
        output_lines=output_lines,
    )

    # Code that might check fileno should not get AttributeError
    with pytest.raises(io.UnsupportedOperation):
        os.isatty(writer.fileno())
```

### Test 4: Stage function using library that checks fileno

This integration test ensures the fix works end-to-end:

```python
def test_run_stage_function_handles_fileno_check() -> None:
    """Stage functions that check stdout.fileno() get proper exception."""

    def stage_checks_fileno() -> None:
        import io
        import sys
        
        # Libraries often check fileno for TTY detection
        try:
            sys.stdout.fileno()
        except io.UnsupportedOperation:
            print("correctly caught UnsupportedOperation")
        except AttributeError:
            raise AssertionError("Should not get AttributeError")

    output_lines: list[tuple[str, bool]] = []
    worker._run_stage_function_with_capture(
        stage_checks_fileno,
        "test_stage",
        mp.Manager().Queue(),  # pyright: ignore[reportArgumentType]
        output_lines,
    )

    assert len(output_lines) == 1
    assert "correctly caught UnsupportedOperation" in output_lines[0][0]
```

## File Changes Summary

| File | Change |
|------|--------|
| `/home/pivot/app/src/pivot/executor/worker.py` | Add `import io` at line 9; Add `fileno()` method after `isatty()` |
| `/home/pivot/app/tests/execution/test_executor_worker.py` | Add 4 new test functions in `_QueueWriter Tests` section |

## Verification Checklist

After implementation, run:

```bash
ruff format .
ruff check .
basedpyright .
pytest tests/execution/test_executor_worker.py -v
pytest tests/ -v  # Full test suite
```

## Future Considerations

The current implementation is minimal and focused on the reported issue. If users encounter issues with other missing methods, consider adding:

1. **`readable() -> bool`**: Return `False` (this is a write-only stream)
2. **`writable() -> bool`**: Return `True`
3. **`seekable() -> bool`**: Return `False`

However, these are less commonly needed than `fileno()` and should only be added if real-world usage demonstrates the need.

---

### Critical Files for Implementation

- `/home/pivot/app/src/pivot/executor/worker.py` - Core file containing `_QueueWriter` class to modify (add `import io` and `fileno()` method)
- `/home/pivot/app/tests/execution/test_executor_worker.py` - Test file with existing `_QueueWriter` tests; add new tests for `fileno()` behavior
- `/home/pivot/app/src/pivot/console.py` - Reference for how project handles `isatty` checks on file-like objects
- `/home/pivot/app/CLAUDE.md` - Project coding standards (import style, docstring format, testing requirements)
- `/home/pivot/app/tests/CLAUDE.md` - Testing conventions (flat functions, parametrization, assertion messages)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by sjawhar**
## Summary

- **Issue #24**: Fix TOCTOU (Time-Of-Check-Time-Of-Use) race condition in `_acquire_execution_lock()` by using atomic rename pattern instead of delete-then-create
- **Issue #26**: Add `fileno()` method to `_QueueWriter` that raises `io.UnsupportedOperation` (matching Python's `io.StringIO` behavior)

### TOCTOU Fix (#24)
- Add `_read_lock_pid()` helper to extract PID from lock file
- Add `_atomic_lock_takeover()` helper using `tempfile.mkstemp()` + `os.replace()` for atomic takeover
- Refactor `_acquire_execution_lock()` to use atomic takeover instead of `unlink()` then retry
- Include fd leak protection with `fd_closed` tracking and `finally` block

### fileno() Method (#26)
- Add `fileno()` method to `_QueueWriter` class
- Raises `io.UnsupportedOperation` per Python's file-like object conventions
- Prevents `AttributeError` when libraries like `tqdm` or `rich` call `fileno()`

## Test plan

- [x] Added `test_queue_writer_fileno_raises_unsupported_operation`
- [x] Added `test_read_lock_pid_returns_pid_for_valid_file`
- [x] Added `test_read_lock_pid_returns_none_for_missing_file`
- [x] Added `test_read_lock_pid_returns_none_for_corrupted_file`
- [x] Added `test_read_lock_pid_returns_none_for_negative_pid`
- [x] Added `test_atomic_lock_takeover_succeeds_on_stale_lock`
- [x] Added `test_atomic_lock_takeover_fails_when_another_process_wins`
- [x] Added `test_atomic_takeover_cleans_temp_on_error`
- [x] Added `test_acquire_lock_retries_after_failed_takeover`
- [x] All 626 tests pass with 91.22% coverage

Closes #24
Closes #26

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by sjawhar — COMMENTED** [review]
## Review Summary

**Recommendation: APPROVE** - This PR correctly addresses both issues (#24 TOCTOU race condition, #26 missing `fileno()` method) with a well-designed solution. The implementation follows the project's coding standards and includes comprehensive test coverage.

## What Works Well

- **TOCTOU fix is sound**: The atomic rename pattern (`tempfile.mkstemp()` + `os.replace()`) eliminates the race window between deleting a stale lock and creating a new one. The verification step after rename ensures only one process wins when multiple processes race to take over.

- **Proper fd leak protection**: The `fd_closed` tracking pattern in `_atomic_lock_takeover()` correctly handles the case where `os.close()` succeeds but `os.replace()` fails, preventing double-close issues. This follows the pattern documented in CLAUDE.md (Critical Discovery #11).

- **Cross-platform compatibility**: Using `os.replace()` instead of `os.rename()` works correctly on both POSIX and Windows systems for atomic overwrites.

- **Clean separation of concerns**: Extracting `_read_lock_pid()` as a separate helper improves readability and makes the code easier to test.

- **`fileno()` implementation follows Python conventions**: Raising `io.UnsupportedOperation` matches `io.StringIO` behavior and inherits from both `OSError` and `ValueError` for backward compatibility with code catching either exception type.

- **Comprehensive test coverage**: Tests cover the happy path, race condition simulation, error handling, and cleanup scenarios.

## Blocking Issues

None identified.

## Important Issues

### 1. **[IMPORTANT]** Potential fd leak in fast path of `_acquire_execution_lock()`

Copilot correctly identified that the fast path (lines 369-372) has a potential fd leak if `os.write()` fails:

```python
fd = os.open(sentinel, os.O_CREAT | os.O_EXCL | os.O_WRONLY)
os.write(fd, f"pid: {os.getpid()}\n".encode())  # If this fails, fd leaks
os.close(fd)
return sentinel
```

While this is unlikely to fail in practice (the string is small and fits in kernel buffer), for consistency with `_atomic_lock_takeover()`, consider wrapping in try-finally:

```python
fd = os.open(sentinel, os.O_CREAT | os.O_EXCL | os.O_WRONLY)
try:
    os.write(fd, f"pid: {os.getpid()}\n".encode())
finally:
    os.close(fd)
return sentinel
```

**Note**: This issue predates the PR (it exists in the original code), so fixing it is optional for this PR. However, since you're modifying this code anyway, it would be a good opportunity to address it.

### 2. **[IMPORTANT]** Missing test for PID 0 edge case

While `_read_lock_pid()` correctly returns `None` for `pid <= 0`, there's no explicit test for PID 0. Add:

```python
def test_read_lock_pid_returns_none_for_zero_pid(worker_env: pathlib.Path) -> None:
    """_read_lock_pid returns None for invalid zero PID."""
    sentinel = worker_env / "test.running"
    sentinel.write_text("pid: 0\n")

    assert worker._read_lock_pid(sentinel) is None
```

### 3. **[IMPORTANT]** Missing coverage for `PermissionError` branch (line 400)

Line 400 (`return True` for `PermissionError`) is not covered by tests. This branch handles the case where the process exists but we don't have permission to signal it. Consider adding a test that mocks `os.kill` to raise `PermissionError`.

## Additional Suggestions

### 1. **[SUGGESTION]** Consider adding test for exhausting retry attempts

The branch at line 389 (raising after `_MAX_LOCK_ATTEMPTS`) is not covered. Consider adding a test that forces all takeover attempts to fail:

```python
def test_acquire_lock_exhausts_attempts_and_fails(
    worker_env: pathlib.Path,
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Lock acquisition fails after exhausting all attempts."""
    sentinel = worker_env / "test_stage.running"
    sentinel.write_text("pid: 999999999\n")
    
    # Always fail takeover
    monkeypatch.setattr(worker, "_atomic_lock_takeover", lambda *_: False)
    
    with pytest.raises(exceptions.StageAlreadyRunningError, match="after 3 attempts"):
        worker._acquire_execution_lock("test_stage", worker_env)
```

### 2. **[SUGGESTION]** Consider making test cleanup more robust

Several tests manually clean up with `sentinel.unlink()` at the end. If an assertion fails mid-test, cleanup won't run. Consider using a fixture or pytest's `tmp_path` more directly to ensure cleanup.

### 3. **[NITPICK]** Import `io` at test module level

In `test_queue_writer_fileno_raises_unsupported_operation()`, `io` is imported inside the test function. Per the project's test guidelines, imports should be at module level:

```python
# At module level
import io

# In test
def test_queue_writer_fileno_raises_unsupported_operation() -> None:
    """_QueueWriter.fileno() raises io.UnsupportedOperation."""
    # ... (no import io here)
    with pytest.raises(io.UnsupportedOperation, match="file descriptor"):
        writer.fileno()
```

## Testing Notes

- All 626 tests pass
- Overall coverage is 91.16% (above 90% threshold)
- `worker.py` coverage is 82.41%, with uncovered lines being:
  - Lines 357-358: finally block fd cleanup (hard to trigger)
  - Line 389: exhaust all retry attempts
  - Line 400: `PermissionError` branch in `_is_process_alive`
- The new tests adequately cover the TOCTOU fix and `fileno()` addition

## Next Steps

1. **Optional**: Fix the fd leak in the fast path (pre-existing issue)
2. **Recommended**: Add test for PID 0 edge case
3. **Consider**: Add test for exhausting retry attempts
4. **Consider**: Move `import io` to module level in test file

Overall, this is a well-implemented fix that correctly addresses both issues with appropriate test coverage. The code follows project conventions and the atomic rename pattern is the right solution for the TOCTOU race condition.

File: src/pivot/executor/worker.py [hunk 1]
```diff
@@ -311,12 +317,54 @@ def _execution_lock(stage_name: str, cache_dir: pathlib.Path) -> Generator[pathl
...
+    fd, tmp_path = tempfile.mkstemp(dir=sentinel.parent, prefix=f".{sentinel.name}.")
+    fd_closed = False
+    try:
+        os.write(fd, f"pid: {my_pid}\n".encode())
+        os.close(fd)
+        fd_closed = True
+        os.replace(tmp_path, sentinel)
+
+        # Verify we still hold the lock (another process may have done the same)
+        if _read_lock_pid(sentinel) == my_pid:
```
Comment (sjawhar) [inline src/pivot/executor/worker.py]: **[SUGGESTION]** Well-designed verification step. This is the key to fixing the TOCTOU race - by re-reading the file after the atomic rename, we ensure that even if multiple processes race to do the rename, only the one whose PID ends up in the file will succeed. Good use of the pattern from CLAUDE.md Critical Discovery #11 for fd tracking.

File: tests/execution/test_executor_worker.py [hunk 2]
```diff
@@ -709,6 +726,143 @@ def test_is_process_alive_returns_true_for_init() -> None:
...
+    sentinel.write_text("garbage content")
+
+    assert worker._read_lock_pid(sentinel) is None
+
+
+def test_read_lock_pid_returns_none_for_negative_pid(worker_env: pathlib.Path) -> None:
+    """_read_lock_pid returns None for invalid negative PID."""
+    sentinel = worker_env / "test.running"
+    sentinel.write_text("pid: -1\n")
+
```
Comment (sjawhar) [inline tests/execution/test_executor_worker.py]: **[IMPORTANT]** Missing test for PID 0 edge case. While `_read_lock_pid()` correctly returns `None` for `pid <= 0`, there's only a test for negative PID (-1). Consider adding a test for PID 0:

```python
def test_read_lock_pid_returns_none_for_zero_pid(worker_env: pathlib.Path) -> None:
    """_read_lock_pid returns None for invalid zero PID."""
    sentinel = worker_env / "test.running"
    sentinel.write_text("pid: 0\n")

    assert worker._read_lock_pid(sentinel) is None
```
